File: apps/api/src/routers/accounts.ts
import { TRPCError } from '@trpc/server';
import { eq, inArray, sql } from 'drizzle-orm';
import { z } from 'zod';
import { db, rawSqlite } from '@finances/db';
import { accounts, transactions } from '@finances/db';
import {
  AccountSchema,
  CreateAccountInput,
  UpdateAccountInput,
  ReorderInput,
} from '@finances/contracts';
import { router, protectedProcedure } from '../trpc/trpc.js';

const IdInput = z.object({ id: z.string().uuid() });

export const accountsRouter = router({
  list: protectedProcedure
    .input(z.object({ includeArchived: z.boolean().default(false) }).optional())
    .output(z.array(AccountSchema))
    .query(({ input }) => {
      const includeArchived = input?.includeArchived ?? false;
      const where = includeArchived ? undefined : eq(accounts.archived, false);
      return db
        .select()
        .from(accounts)
        .where(where)
        .orderBy(accounts.sortOrder)
        .all();
    }),

  byId: protectedProcedure
    .input(IdInput)
    .output(AccountSchema)
    .query(({ input }) => {
      const row = db.select().from(accounts).where(eq(accounts.id, input.id)).get();
      if (!row) {
        throw new TRPCError({ code: 'NOT_FOUND', message: 'Compte no trobat' });
      }
      return row;
    }),

  create: protectedProcedure
    .input(CreateAccountInput)
    .output(AccountSchema)
    .mutation(({ input }) => {
      const id = crypto.randomUUID();
      const now = new Date().toISOString();
      const maxRow = db
        .select({ s: accounts.sortOrder })
        .from(accounts)
        .orderBy(accounts.sortOrder)
        .all()
        .at(-1);
      db.insert(accounts)
        .values({
          id,
          name: input.name,
          type: input.type,
          currency: input.currency ?? 'EUR',
          color: input.color,
          icon: input.icon,
          initialBalance: input.initialBalance,
          sortOrder: (maxRow?.s ?? -1) + 1,
          archived: false,
          createdAt: now,
          updatedAt: now,
        })
        .run();
      const row = db.select().from(accounts).where(eq(accounts.id, id)).get();
      if (!row) throw new Error('Failed to create account');
      return row;
    }),

  update: protectedProcedure
    .input(UpdateAccountInput)
    .output(AccountSchema)
    .mutation(({ input }) => {
      const { id, ...patch } = input;
      const existing = db.select().from(accounts).where(eq(accounts.id, id)).get();
      if (!existing) {
        throw new TRPCError({ code: 'NOT_FOUND', message: 'Compte no trobat' });
      }
      const now = new Date().toISOString();
      db.update(accounts)
        .set({ ...patch, updatedAt: now })
        .where(eq(accounts.id, id))
        .run();
      const row = db.select().from(accounts).where(eq(accounts.id, id)).get();
      if (!row) throw new Error('Failed to update account');
      return row;
    }),

  archive: protectedProcedure
    .input(IdInput)
    .output(AccountSchema)
    .mutation(({ input }) => {
      const existing = db.select().from(accounts).where(eq(accounts.id, input.id)).get();
      if (!existing) {
        throw new TRPCError({ code: 'NOT_FOUND', message: 'Compte no trobat' });
      }
      const now = new Date().toISOString();
      db.update(accounts)
        .set({ archived: true, updatedAt: now })
        .where(eq(accounts.id, input.id))
        .run();
      const row = db.select().from(accounts).where(eq(accounts.id, input.id)).get();
      if (!row) throw new Error('Failed to archive account');
      return row;
    }),

  /**
   * Hard-delete an account. Cascades to its transactions (both legs) so we
   * don't leave orphan rows behind the FK in `transactions.account_id`. Wrapped
   * in a single better-sqlite3 transaction for atomicity.
   *
   * For accounts involved in a transfer (transferAccountId), we also remove
   * the counter-leg transaction so the books stay consistent.
   */
  delete: protectedProcedure
    .input(IdInput)
    .output(z.object({ id: z.string().uuid(), deletedTransactions: z.number().int() }))
    .mutation(({ input }) => {
      const existing = db.select().from(accounts).where(eq(accounts.id, input.id)).get();
      if (!existing) {
        throw new TRPCError({ code: 'NOT_FOUND', message: 'Compte no trobat' });
      }
      const txCount = db
        .select({ id: transactions.id })
        .from(transactions)
        .where(eq(transactions.accountId, input.id))
        .all();
      const counterLegs = db
        .select({ id: transactions.id })
        .from(transactions)
        .where(eq(transactions.transferAccountId, input.id))
        .all();

      const deleteAll = rawSqlite.transaction(() => {
        if (txCount.length > 0) {
          db.delete(transactions).where(eq(transactions.accountId, input.id)).run();
        }
        if (counterLegs.length > 0) {
          db.delete(transactions).where(eq(transactions.transferAccountId, input.id)).run();
        }
        db.delete(accounts).where(eq(accounts.id, input.id)).run();
      });
      deleteAll();

      return { id: input.id, deletedTransactions: txCount.length + counterLegs.length };
    }),

  reorder: protectedProcedure
    .input(ReorderInput)
    .output(z.object({ count: z.number().int() }))
    .mutation(({ input }) => {
      const now = new Date().toISOString();
      const updates = input.ids.map((id, idx) =>
        db
          .update(accounts)
          .set({ sortOrder: idx, updatedAt: now })
          .where(eq(accounts.id, id))
          .run(),
      );
      if (updates.length !== input.ids.length) {
        throw new TRPCError({
          code: 'INTERNAL_SERVER_ERROR',
          message: "Alguns IDs no s'han trobat",
        });
      }
      const matched = db
        .select({ id: accounts.id })
        .from(accounts)
        .where(inArray(accounts.id, input.ids))
        .all();
      if (matched.length !== input.ids.length) {
        throw new TRPCError({
          code: 'NOT_FOUND',
          message: 'Algun dels IDs no existeix',
        });
      }
      return { count: input.ids.length };
    }),

  /**
   * Current balance per account: initial_balance + Σ(income) − Σ(expense).
   * Transfer entries net to zero (one +leg, one −leg) so they're included as-is.
   * Used by /accounts list to render the running balance.
   */
  balances: protectedProcedure
    .output(z.array(z.object({ accountId: z.string().uuid(), balanceCents: z.number().int() })))
    .query(() => {
      // Exclude archived accounts — they should not contribute to the running
      // total on the dashboard / accounts list. (See Bug 1: archive in the UI
      // used to be invisible because list ignored `archived` and the running
      // total kept counting them.)
      const accs = db
        .select()
        .from(accounts)
        .where(eq(accounts.archived, false))
        .all();
      const txRows = db
        .select({
          accountId: transactions.accountId,
          income: sql<number>`COALESCE(SUM(CASE WHEN ${transactions.kind} = 'income' THEN ${transactions.amount} ELSE 0 END), 0)`,
          expense: sql<number>`COALESCE(SUM(CASE WHEN ${transactions.kind} = 'expense' THEN ${transactions.amount} ELSE 0 END), 0)`,
        })
        .from(transactions)
        .groupBy(transactions.accountId)
        .all();

      const txMap = new Map<string, { income: number; expense: number }>();
      for (const row of txRows) txMap.set(row.accountId, { income: row.income, expense: row.expense });

      return accs.map((a) => {
        const totals = txMap.get(a.id) ?? { income: 0, expense: 0 };
        const balanceCents = a.initialBalance + totals.income - totals.expense;
        return { accountId: a.id, balanceCents };
      });
    }),
});

File: backend/internal/db/accounts.go
package db

import (
	"database/sql"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/mgurt/finances/internal/models"
)

// CreateAccountReq is the input for creating an account.
type CreateAccountReq struct {
	Name           string
	Type           string
	Color          string
	Icon           string
	InitialBalance int
	Currency       string
}

// UpdateAccountReq is the input for updating an account.
type UpdateAccountReq struct {
	Name  string
	Type  string
	Color string
	Icon  string
}

// AccountsStore provides query methods for accounts.
type AccountsStore struct {
	*Store
}

// List returns all accounts, optionally filtered to exclude archived.
func (a *AccountsStore) List(includeArchived bool) ([]models.Account, error) {
	query := "SELECT id, name, type, currency, color, icon, initial_balance, sort_order, archived, created_at, updated_at FROM accounts"
	if !includeArchived {
		query += " WHERE archived = 0"
	}
	query += " ORDER BY sort_order ASC, created_at ASC"

	rows, err := a.DB.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var accounts []models.Account
	for rows.Next() {
		var acc models.Account
		err := rows.Scan(&acc.ID, &acc.Name, &acc.Type, &acc.Currency, &acc.Color, &acc.Icon, &acc.InitialBalance, &acc.SortOrder, &acc.Archived, &acc.CreatedAt, &acc.UpdatedAt)
		if err != nil {
			return nil, err
		}
		accounts = append(accounts, acc)
	}
	return accounts, rows.Err()
}

// ByID returns a single account by ID.
func (a *AccountsStore) ByID(id string) (*models.Account, error) {
	query := "SELECT id, name, type, currency, color, icon, initial_balance, sort_order, archived, created_at, updated_at FROM accounts WHERE id = ?"
	var acc models.Account
	err := a.DB.QueryRow(query, id).Scan(&acc.ID, &acc.Name, &acc.Type, &acc.Currency, &acc.Color, &acc.Icon, &acc.InitialBalance, &acc.SortOrder, &acc.Archived, &acc.CreatedAt, &acc.UpdatedAt)
	if err == sql.ErrNoRows {
		return nil, fmt.Errorf("account not found")
	}
	if err != nil {
		return nil, err
	}
	return &acc, nil
}

// Create inserts a new account and returns it.
func (a *AccountsStore) Create(req CreateAccountReq) (*models.Account, error) {
	id := uuid.New().String()
	now := time.Now().UTC().Format(time.RFC3339)
	currency := req.Currency
	if currency == "" {
		currency = "EUR"
	}
	color := req.Color
	if color == "" {
		color = "#6366F1"
	}
	icon := req.Icon
	if icon == "" {
		icon = "wallet"
	}

	_, err := a.DB.Exec(`
		INSERT INTO accounts (id, name, type, currency, color, icon, initial_balance, sort_order, archived, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, 0, 0, ?, ?)`,
		id, req.Name, req.Type, currency, color, icon, req.InitialBalance, now, now)
	if err != nil {
		return nil, err
	}

	return a.ByID(id)
}

// Update modifies an existing account and returns the updated account.
func (a *AccountsStore) Update(id string, req UpdateAccountReq) (*models.Account, error) {
	existing, err := a.ByID(id)
	if err != nil {
		return nil, err
	}

	name := req.Name
	if name == "" {
		name = existing.Name
	}
	accountType := req.Type
	if accountType == "" {
		accountType = existing.Type
	}
	color := req.Color
	if color == "" {
		color = existing.Color
	}
	icon := req.Icon
	if icon == "" {
		icon = existing.Icon
	}
	now := time.Now().UTC().Format(time.RFC3339)

	_, err = a.DB.Exec("UPDATE accounts SET name = ?, type = ?, color = ?, icon = ?, updated_at = ? WHERE id = ?",
		name, accountType, color, icon, now, id)
	if err != nil {
		return nil, err
	}

	return a.ByID(id)
}

// Archive sets archived=1 for the given account.
func (a *AccountsStore) Archive(id string) error {
	now := time.Now().UTC().Format(time.RFC3339)
	result, err := a.DB.Exec("UPDATE accounts SET archived = 1, updated_at = ? WHERE id = ?", now, id)
	if err != nil {
		return err
	}
	rows, err := result.RowsAffected()
	if err != nil {
		return err
	}
	if rows == 0 {
		return fmt.Errorf("account not found")
	}
	return nil
}

// Delete removes an account and all its transaction legs.
// Uses a transaction for atomicity. The account_id FK has ON DELETE CASCADE,
// and we explicitly delete transfer_account_id legs to ensure they're counted.
// Returns the count of transactions deleted.
func (a *AccountsStore) Delete(id string) (int, error) {
	_, err := a.ByID(id)
	if err != nil {
		return 0, err
	}

	// Count all transactions that will be deleted
	var txCount int
	err = a.DB.QueryRow(`
		SELECT COUNT(*) FROM transactions
		WHERE account_id = ? OR transfer_account_id = ?`, id, id).Scan(&txCount)
	if err != nil {
		return 0, err
	}

	// Use transaction for atomicity
	tx, err := a.DB.Begin()
	if err != nil {
		return 0, err
	}
	defer tx.Rollback()

	// Delete transfer legs explicitly (transfer_account_id FK doesn't have CASCADE)
	_, err = tx.Exec("DELETE FROM transactions WHERE transfer_account_id = ?", id)
	if err != nil {
		return 0, err
	}

	// Delete account (account_id FK has CASCADE, so account_id legs are cascade-deleted)
	_, err = tx.Exec("DELETE FROM accounts WHERE id = ?", id)
	if err != nil {
		return 0, err
	}

	if err := tx.Commit(); err != nil {
		return 0, err
	}

	return txCount, nil
}

// Reorder updates sort_order for accounts based on the provided ID order.
func (a *AccountsStore) Reorder(order []string) error {
	if len(order) == 0 {
		return nil
	}

	tx, err := a.DB.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	for i, id := range order {
		_, err := tx.Exec("UPDATE accounts SET sort_order = ? WHERE id = ?", i, id)
		if err != nil {
			return err
		}
	}

	return tx.Commit()
}

// Balances returns all accounts with their computed balances
// (initial_balance + sum of transaction amounts).
func (a *AccountsStore) Balances() ([]models.AccountWithBalance, error) {
	query := `
		SELECT
			a.id, a.name, a.type, a.currency, a.color, a.icon, a.initial_balance,
			a.sort_order, a.archived, a.created_at, a.updated_at,
			COALESCE(SUM(t.amount), 0) + a.initial_balance AS balance
		FROM accounts a
		LEFT JOIN transactions t ON t.account_id = a.id
		GROUP BY a.id
		ORDER BY a.sort_order ASC, a.created_at ASC`

	rows, err := a.DB.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []models.AccountWithBalance
	for rows.Next() {
		var awb models.AccountWithBalance
		err := rows.Scan(
			&awb.ID, &awb.Name, &awb.Type, &awb.Currency, &awb.Color, &awb.Icon,
			&awb.InitialBalance, &awb.SortOrder, &awb.Archived, &awb.CreatedAt, &awb.UpdatedAt,
			&awb.Balance,
		)
		if err != nil {
			return nil, err
		}
		results = append(results, awb)
	}
	return results, rows.Err()
}